lib/data.js

```diff
@@ -24,47 +24,40 @@ api.common.copy = (
   return result;
 */
 
-api.common.clone = (
-  // Clone dataset (clone objects to new array)
-  ds
-) => cloneArray(ds);
+api.common.clone = (obj) => {
+  if (typeof obj !== 'object' || obj === null) return obj;
+  return Array.isArray(obj) ? cloneArray(obj) : cloneObject(obj);
+};
 
-function cloneObject(obj) {
-  const result = {};
-  const keys = Object.keys(obj);
-  const len = keys.length;
-  let i, key, val, type;
-  for (i = 0; i < len; i++) {
-    key = keys[i];
-    val = obj[key];
-    type = typeof(val);
-    if (type === 'object') {
-      result[key] = (
-        Array.isArray(val) ? cloneArray(val) : cloneObject(val)
-      );
+function cloneArray(arr) {
+  const size = arr.length;
+  const array = new Array(size);
+  let i, val;
+  for (i = 0; i < size; i++) {
+    val = arr[i];
+    if (typeof val !== 'object' || val === null) {
+      array[i] = val;
     } else {
-      result[key] = val;
+      array[i] = Array.isArray(val) ? cloneArray(val) : cloneObject(val);
```

Comment:
Why not use `api.common.clone()` here? It already contains this type checking logic.